src/sources/socket.rs

```diff
@@ -0,0 +1,233 @@
+use super::util::TcpSource;
+use crate::{
+    event::{self, Event},
+    sources::tcp::{RawTcpSource, TcpConfig},
+    sources::udp::{udp, UdpConfig},
+    topology::config::{DataType, GlobalOptions, SourceConfig, SourceDescription},
+};
+use futures::sync::mpsc;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+// TODO: add back when serde-rs/serde#1358 is addressed
+// #[serde(deny_unknown_fields)]
+pub struct SocketConfig {
+    #[serde(flatten)]
+    pub mode: Mode,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(tag = "mode", rename_all = "snake_case")]
+pub enum Mode {
+    Tcp(TcpConfig),
+    Udp(UdpConfig),
+}
+
+impl SocketConfig {
+    pub fn new(mode: Mode) -> Self {
+        Self { mode }
+    }
+}
```

Comment:
The precedent for this kind of constructor would be to write an `impl From` for each of `TcpConfig` and `UdpConfig` which would allow using `.into()` on a `TcpConfig` to create this without needing to explicitly use `Mode`. See `src/event/mod.rs`.

Reply:
Done!  Thanks that's definitely much nicer looking now.